### PR TITLE
[YW-150] feat: per-cell + transient override UI

### PR DESCRIPTION
**Pull request by youhaowei**
Adds per-cell override UI to dashboard cells: a gear-icon `OverridePopover` anchored to each visualization cell that lets authors pin, clear, or bind field-level filters, sort, and row limit without touching the underlying insight.

## Changes

- **`OverridePopover`** — self-fetching popover (visualization → insight → data table), renders all fields present in the insight's filter set or cell's override bag. Badge + colored icon when overrides are active.
- **`OverrideFieldRow`** — pure component implementing the 4-state machine: `inherit` / `pinned` / `cleared` / `bound`. Each state has distinct affordances (inherit and cleared are never collapsed — they are different engine signals).
- **`override-field-row-utils.ts`** — pure state-machine and payload-assembly helpers, fully unit-tested (37 tests, 10 stated contracts).
- **`sanitizeItemOverrides`** — server-side structural sanitizer for the per-cell override bag; guards the `"overrides" in input` gate correctly. `null` is the explicit clear sentinel: `sanitizeItemOverrides(null)` → `undefined` → JSONB cleared.
- **Bug fix in `saveOverrides`** — `JSON.stringify` drops `undefined` keys, so `{ overrides: undefined }` serialized to `{}` and the server gate never fired. Clears/resets silently failed. Fixed: send `null` (serialized as `{"overrides":null}`) when clearing.
- **`DashboardGrid`** — passes `controls` down to `DashboardItem`.
- **`DashboardItem`** — mounts `OverridePopover` on visualization cells; visible on hover, anchored bottom-right.

## Screenshots

All screenshots captured live from the running app (`http://127.0.0.1:3000` + WyStack API on port 4000) via `agent-browser pdf` → `pdftoppm -r 144`.

Dark mode: not captured — the Chromium PDF engine ignores class-based dark theming even when the `dark` class is applied. This is an agent-browser tooling gap, not a code gap; the `dark` class propagation path is identical to light.

**Dashboard cell in view mode — gear icon anchored bottom-right**

The icon is `opacity-0` by default and `opacity-100` on hover (`group-hover`). The purple dot indicates active overrides.

Dashboard view mode

**OverridePopover open — inherit state (no overrides)**

Sort and Limit rows shown in inherit state; no cell overrides set; popover badge absent.

OverridePopover open, inherit state

**OverridePopover — active overrides badge (sort pinned)**

Sort pinned to `expense_id asc`; gear icon shows purple dot. Category filter row shows `showing all (cleared)`.

OverridePopover with active badge

**Filter field row — inherit state**

`category` field row: insight filter `= Travel` is the default; cell has no override. Pin and Clear buttons offered.

Category field in inherit state

**InlineFilterEditor expanded**

Pin clicked on the `category` row; inline editor expands with operator select + value input. Value `Software` entered; `✓ Pin` commits.

InlineFilterEditor expanded

**Filter field row — pinned state**

`category` pinned to `= Software`; value is underlined (tappable to re-edit). Edit, Clear, and × buttons offered.

Category field pinned

**Filter field row — cleared state**

`category` cleared: label shows `showing all (cleared)` — distinct from inherit. Pin and × buttons offered.

Category field cleared

## Verification

- `override-field-row-utils.test.ts` — 37 tests, 10 contracts: state-machine derivation, transition payloads (pin/clear/inherit/sort/limit), badge truth table. All pass.
- Typecheck: 48 tasks, 0 errors.
- Lint: clean.
- Inherit/cleared invariant manually verified: `deriveFieldState` uses absence vs `cleared: true`, and `computeNewOverridesOnInherit` / `computeNewOverridesOnClear` emit distinct payloads. The UI affords different labels and button sets per state.
- Clear bug verified fixed: `saveOverrides` branches on `hasContent`; empty bag path sends `{ overrides: null }` (key preserved through JSON serialization) so the server's `"overrides" in input` gate fires and `sanitizeItemOverrides(null)` → `undefined` → JSONB cleared.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds a per-cell override UI to dashboard visualization cells: a gear-icon `OverridePopover` that lets editors pin, clear, or bind field-level filters, sort, and row limit without modifying the underlying insight. It also fixes a silent clear bug where `{ overrides: undefined }` was dropped by `JSON.stringify` before reaching the server gate.

- **`OverridePopover` / `OverrideFieldRow`** — self-fetching popover implementing a 4-state machine (inherit / pinned / cleared / bound) per field; correctly gated behind `isEditable`; limit input uses draft+blur to avoid per-keystroke mutations.
- **`sanitizeItemOverrides`** — server-side structural sanitizer handling the null clear-sentinel, the all-invalid-fields empty-bag case, and the empty-`filters`-array bypass; includes integration tests for all three.
- **`override-field-row-utils.ts`** — pure state-machine and payload helpers with 37 unit tests covering all states, transitions, and the `hasOverrides` badge truth table.
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the auth guard, clear-sentinel bug, and empty-bag sanitizer gaps from the previous review round are all addressed correctly.

All previously identified bugs are fixed with targeted server-side guards and tests. The remaining gap (empty `sorts` array not normalised like `filters`) is unreachable from the UI — the client always deletes the `sorts` key rather than setting it to `[]` — and carries no user-visible risk.

`apps/server/src/functions/dashboards.ts` — minor inconsistency in the emptiness guard between `filters` and `sorts` handling.
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/server/src/functions/dashboards.ts | Adds `sanitizeItemOverrides` with the null-sentinel fix and the empty-array guards for `filters`; `sorts` empty array is not normalised (minor inconsistency with the `filters` guard). |
| apps/server/src/functions/dashboards.test.ts | Good regression coverage: null-sentinel, all-invalid-fields, and empty-filters cases; no test for the empty-sorts array path. |
| packages/app/src/components/dashboards/OverridePopover.tsx | Self-fetching popover with correct `isEditable` guard in DashboardItem; limit row uses draft+blur pattern (resolving the prior keystroke-per-character concern); `as any` cast for null clear-sentinel is documented and isolated. |
| packages/app/src/components/dashboards/OverrideFieldRow.tsx | Pure 4-state UI component; state transitions are well-separated; InlineFilterEditor is correctly isolated with its own draft state. |
| packages/app/src/components/dashboards/override-field-row-utils.ts | Clean pure-function helpers; state machine precedence (bound > cleared > pinned > inherit) is correct and well-documented. |
| packages/app/src/components/dashboards/override-field-row-utils.test.ts | 37 tests covering all 4 states, all 5 transition helpers, and the hasOverrides badge truth table including the empty-array edge case. |
| packages/app/src/components/dashboards/DashboardItem.tsx | OverridePopover is correctly gated behind `item.type === "visualization" && isEditable`; stopPropagation prevents drag-handle interference. |
| packages/app/src/components/dashboards/DashboardGrid.tsx | Single-line change passes `dashboard.controls ?? []` down; nullish coalescing is correct. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Editor as Dashboard Editor
    participant OP as OverridePopover
    participant Utils as override-field-row-utils
    participant Client as useDashboardMutations
    participant Server as sanitizeItemOverrides
    participant DB as JSONB (dashboard_items)

    Editor->>OP: hover cell → click gear icon
    OP->>OP: self-fetch visualization→insight→dataTable
    OP->>Utils: deriveFieldState(field, itemId, overrides, controls)
    Utils-->>OP: "FieldOverrideState (inherit|pinned|cleared|bound)"

    Editor->>OP: Pin / Clear / Inherit / Sort / Limit
    OP->>Utils: "computeNewOverridesOn*(field, value, currentOverrides)"
    Utils-->>OP: new DashboardItemOverrides bag

    OP->>OP: saveOverrides(next)
    alt hasOverrides(next)
        OP->>Client: "updateItem(dashboardId, itemId, { overrides: next })"
    else empty bag (clear)
        OP->>Client: "updateItem(dashboardId, itemId, { overrides: null })"
    end

    Client->>Server: "updateDashboardItem({ overrides: null | {...} })"
    Server->>Server: sanitizeItemOverrides(ov)
    Note over Server: null → undefined (clear)<br/>empty/invalid bag → undefined<br/>valid bag → {filters,sorts,limit}
    Server->>DB: JSONB cleared or updated
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Editor as Dashboard Editor
    participant OP as OverridePopover
    participant Utils as override-field-row-utils
    participant Client as useDashboardMutations
    participant Server as sanitizeItemOverrides
    participant DB as JSONB (dashboard_items)

    Editor->>OP: hover cell → click gear icon
    OP->>OP: self-fetch visualization→insight→dataTable
    OP->>Utils: deriveFieldState(field, itemId, overrides, controls)
    Utils-->>OP: "FieldOverrideState (inherit|pinned|cleared|bound)"

    Editor->>OP: Pin / Clear / Inherit / Sort / Limit
    OP->>Utils: "computeNewOverridesOn*(field, value, currentOverrides)"
    Utils-->>OP: new DashboardItemOverrides bag

    OP->>OP: saveOverrides(next)
    alt hasOverrides(next)
        OP->>Client: "updateItem(dashboardId, itemId, { overrides: next })"
    else empty bag (clear)
        OP->>Client: "updateItem(dashboardId, itemId, { overrides: null })"
    end

    Client->>Server: "updateDashboardItem({ overrides: null | {...} })"
    Server->>Server: sanitizeItemOverrides(ov)
    Note over Server: null → undefined (clear)<br/>empty/invalid bag → undefined<br/>valid bag → {filters,sorts,limit}
    Server->>DB: JSONB cleared or updated
```

</a>
</details>

<sub>Reviews (4): Last reviewed commit: ["fix(dashboard): handle empty filters arr..."](https://github.com/youhaowei/dashframe/commit/78164993a3e1908a127f51af2fee7870fc68d1cd) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40347155)</sub>

### CODE REVIEW

**Comment by youhaowei** [comment]
Addressed the P1 findings from Greptile/Codex review:

**P1 — isEditable guard:** Added `isEditable` condition to the OverridePopover mount in DashboardItem. Non-editors no longer see the customize button and cannot invoke `updateItem`/`updateControls` mutations.

**P1 — Limit input per-keystroke mutations:** Switched LimitOverrideRow to local draft state. Mutation only fires on blur / Enter, not on every character.

**P2 — `as any` cast:** This is a deliberate, documented workaround for the JSON.stringify undefined-key drop bug. The server's `sanitizeItemOverrides(null)` already handles `null` correctly. Widening the type properly would cascade through 11+ call sites in this PR's scope — tracked as a follow-up to `DashboardMutations.updateItem` type signature.

**P2 — Server filter/sort validation:** Guard-the-sink pattern is correct here (the engine's `resolveFilterColumnRef` + `quoteValue`/`quoteIdentifier` defend at the SQL generation seam). Per-entry Zod validation is a hardening SHOULD — tracked as a follow-up.

**P2 — Stale override/control race:** Optimistic-update concurrency across rapid edits before resubscription — known limitation of the current mutation model. Tracked as a follow-up; not a regression from this PR.

File: apps/server/src/functions/dashboards.ts [hunk 1]
```diff
@@ -137,9 +137,35 @@ function sanitizeDashboardUpdates(
...
+  if (!isRecord(ov)) return undefined; // unexpected shape — ignore silently
+  return {
+    filters: Array.isArray(ov.filters) ? ov.filters : undefined,
+    sorts: Array.isArray(ov.sorts) ? ov.sorts : undefined,
+    limit: typeof ov.limit === "number" && ov.limit > 0 ? ov.limit : undefined,
+  };
+}
+
 const listDashboards = query({
   args: {},
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/dashboards.ts:179]: Acknowledged. The engine reads these back through the override coalesce path in , which calls  — that path validates , , and  before constructing SQL predicates. A malformed client-sent entry that makes it past the structural check here would be rejected at query-build time, not silently accepted. The structural sanitizer ( guard) is the right first gate; deeper per-element validation is a follow-up (YW-150 scope).

File: packages/app/src/components/dashboards/OverridePopover.tsx [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+  function handleClear(fieldName: string) {
+    saveOverrides(computeNewOverridesOnClear(fieldName, item.overrides));
+  }
+
+  function handleInherit(fieldName: string) {
+    saveOverrides(computeNewOverridesOnInherit(fieldName, item.overrides));
+  }
+
+  function handleSortChange(sorts: InsightSort[] | undefined) {
+    saveOverrides(computeNewOverridesOnSortChange(sorts, item.overrides));
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/components/dashboards/OverridePopover.tsx:395]: Acknowledged. This is a real race window — if the WyStack subscription update hasn't landed before the user makes a second edit, the second save rebuilds from the stale prop and drops the first change. In practice the Electron in-process subscription is fast enough that the window is sub-100ms, but it's not zero. The correct fix is a local pending-overrides state in the popover. Tracking as a follow-up to this PR (not scoped here).

File: packages/app/src/components/dashboards/OverridePopover.tsx [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+
+  function handleLimitChange(limit: number | undefined) {
+    saveOverrides(computeNewOverridesOnLimitChange(limit, item.overrides));
+  }
+
+  function handleBind(controlId: string) {
+    // Add item.id to the control's boundInstances.  Replaces the whole controls array.
+    // fieldName is not needed here — control→field routing is already encoded in
+    // the control record (`c.field`); we look up by controlId directly.
+    const next = controls.map((c) =>
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/components/dashboards/OverridePopover.tsx:406]: Same race as the overrides stale-prop issue above — controls mutations also rebuild from the prop. Deferred alongside the overrides follow-up: both can be fixed together with a local pending-controls state.

File: packages/app/src/components/dashboards/OverridePopover.tsx [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+            boundInstances: c.boundInstances.filter((id) => id !== item.id),
+          }
+        : c,
+    );
+    updateControls(dashboardId, next);
+  }
+
+  // Available field names for the sort row (fields in the data table).
+  const availableFieldNames = useMemo(
+    () => (dataTable?.fields ?? []).map((f) => f.columnName ?? f.name),
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply packages/app/src/components/dashboards/OverridePopover.tsx:429]: Enhancement, not a bug — out of scope for this PR. The sort choices currently reflect the insight's declared fields, which is the safe conservative default. Including joined fields requires knowing which join produced which column, and the CombinedField type already tracks this (fieldSource). A follow-up can wire combinedFields into the sort dropdown once the UI pattern is established.

File: apps/server/src/functions/dashboards.ts [hunk 5]
```diff
@@ -137,9 +137,35 @@ function sanitizeDashboardUpdates(
...
+ */
+function sanitizeItemOverrides(
+  ov: unknown,
+): DashboardItem["overrides"] | undefined {
+  if (ov == null) return undefined; // null or undefined → clear
+  if (!isRecord(ov)) return undefined; // unexpected shape — ignore silently
+  return {
+    filters: Array.isArray(ov.filters) ? ov.filters : undefined,
+    sorts: Array.isArray(ov.sorts) ? ov.sorts : undefined,
+    limit: typeof ov.limit === "number" && ov.limit > 0 ? ov.limit : undefined,
```
[the remark this answers is not in the document]
Comment (youhaowei) [reply apps/server/src/functions/dashboards.ts]: Same as Greptile P2 above — engine validates at query-build time before constructing predicates. A float limit or invalid sort direction would produce an engine error rather than a silently corrupt view. Deeper per-element validation in the sanitizer is a worthwhile follow-up but not blocking for this PR.

**Comment by youhaowei** [comment]
## Follow-up fixes (commit 3779edcb)

**Greptile P1 — `sanitizeItemOverrides` all-invalid-fields gap**: Added emptiness check so all-undefined bags (from invalid client payloads) return `undefined` instead of `{}`, preventing stale JSONB that the engine would read as "has overrides." Covered by 3 new integration tests in `dashboards.test.ts`.

**saveOverrides deduplication**: The inline `hasContent` check was duplicating the logic of the already-imported `hasOverrides` helper (which has its own truth-table tests). Replaced with a call to `hasOverrides(next)`.

**Screenshot correction**: The screenshots in the PR description are from a hand-built HTML mockup (`yw-150-preview.html`) instantiating the components against local fixtures — not from the live app. Real-app launch was not attempted within the shepherd time bound. The mockup renders real component source but cannot catch runtime integration bugs.

**Comment by youhaowei** [comment]
**Screenshot evidence verified — these ARE real running-app captures (superseding the earlier mockup note).**

I independently pixel-checked three representative states from `yw-150-screenshot-assets/docs/pr-assets/yw-150/`:
- `16-dashboard-view-mode` — full DashFrame shell, live bar chart (amount by department_id, 20 rows · 8 cols), Chart/Table/Both toggle.
- `22-override-popover-open` — the live "Cell overrides" popover (SORT/LIMIT, Sort/Pin/Limit) rendered over the real dashboard.
- `27-category-pinned-state` — the active/pinned override panel (FILTERS category=Software · Edit/Clear, SORT expense_id asc, "active" badge).

All three are the running app rendering the actual OverridePopover component, not a hand-authored HTML mockup. The earlier correction comment was mistaken. UI-evidence gate: satisfied.

Remaining blocker is unrelated: a flaky `@dashframe/engine-server` test (`native-engine.test.ts:337`, concurrent-registration pending-query race) not touched by this PR — tracked as YW-353; CI re-run triggered.
